server/pubsub/redis_query_results.go

```diff
@@ -15,17 +16,127 @@ import (
 
 type redisQueryResults struct {
 	// connection pool
-	pool             *redisc.Cluster
+	pool             fleet.RedisPool
 	duplicateResults bool
 }
 
 var _ fleet.QueryResultStore = &redisQueryResults{}
 
+// this is an adapter type to implement the same Stats method as for
+// redisc.Cluster, so both can satisfy the same interface.
+type standalonePool struct {
+	*redis.Pool
+	addr string
+}
+
+func (p *standalonePool) Stats() map[string]redis.PoolStats {
+	return map[string]redis.PoolStats{
+		p.addr: p.Pool.Stats(),
+	}
+}
+
 // NewRedisPool creates a Redis connection pool using the provided server
 // address, password and database.
-func NewRedisPool(server, password string, database int, useTLS bool) (*redisc.Cluster, error) {
-	// Create the Cluster
-	cluster := &redisc.Cluster{
+func NewRedisPool(server, password string, database int, useTLS bool) (fleet.RedisPool, error) {
+	cluster := newCluster(server, password, database, useTLS)
+	if err := cluster.Refresh(); err != nil {
+		if isClusterDisabled(err) || isClusterCommandUnknown(err) {
+			// not a Redis Cluster setup, use a standalone Redis pool
+			pool, _ := cluster.CreatePool(server)
+			cluster.Close()
+			return &standalonePool{pool, server}, nil
+		}
+		return nil, errors.Wrap(err, "refresh cluster")
+	}
+
+	return cluster, nil
+}
+
+// SplitRedisKeysBySlot takes a list of redis keys and groups them by hash slot
+// so that keys in a given group are guaranteed to hash to the same slot, making
+// them safe to run e.g. in a pipeline on the same connection or as part of a
+// multi-key command in a Redis Cluster setup. When using standalone Redis, it
+// simply returns all keys in the same group (i.e. the top-level slice has a
+// length of 1).
+func SplitRedisKeysBySlot(pool fleet.RedisPool, keys ...string) [][]string {
+	if _, isCluster := pool.(*redisc.Cluster); isCluster {
+		return redisc.SplitBySlot(keys...)
+	}
+	return [][]string{keys}
+}
+
+// EachRedisNode calls fn for each node in the redis cluster, with a connection
+// to that node, until all nodes have been visited. The connection is
+// automatically closed after the call. If fn returns an error, the iteration
+// of nodes stops and EachRedisNode returns that error. For standalone redis,
+// fn is called only once.
+func EachRedisNode(pool fleet.RedisPool, fn func(conn redis.Conn) error) error {
+	// TODO: ideally, NewRedisPool, EachRedisNode and other helper functions
+	// would live in a datastore/redis package or something like that? This is
+	// not related to pubsub specifically.
```

Comment:
+1!

Reply:
Do you think `datastore/redis` specifically makes sense? I was wondering if that was confusing because the redis pool is otherwise unrelated to the `fleet.Datastore` interface. Maybe I'm overthinking this :D